// fb-messenger-bot-api/dist/util/utils.js
/* eslint-disable semi */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.requestOptions = exports.deepCopyPayload = exports.sendMessage = undefined;

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sendMessage = function sendMessage(options, requestData, cb) {
  options.qs.access_token = requestData.token;
  if (requestData.hasOwnProperty('proxy')) {
    options.proxy = requestData.proxy;
  }

  if (typeof cb !== 'function') {
    return new Promise(function (resolve, reject) {
      (0, _request2.default)(options, function (err, res, body) {
        if (err) {
          reject(err);
        } else {
          if (!(body && body.constructor === Object)) {
            var bodyObject = JSON.parse(body);
            if (bodyObject.error) {
              reject(bodyObject.error.message);
            } else {
              resolve(bodyObject);
            }
          } else {
            resolve(body);
          }
        }
      });
    });
  } else {
    (0, _request2.default)(options, function (err, res, body) {
      if (err) return cb(err);
      if (body.error) return cb(body.error);
      cb(null, body);
    });
  }
};

var deepCopyPayload = function deepCopyPayload(payloadType) {
  return JSON.parse(JSON.stringify(payloadType));
};

var requestOptions = {
  url: 'https://graph.facebook.com/v2.10/',
  qs: {
    access_token: undefined
  },
  method: undefined
};

exports.sendMessage = sendMessage;
exports.deepCopyPayload = deepCopyPayload;
exports.requestOptions = requestOptions;
